### DISCUSSION TITLE
How to import custom python files in `app.py`

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/briefcase#1124] [Q&A] [answered]
If I have the following structure inside `src` directory:
|-->`app.py`
|-->`__init__.py`
|-->`__main__.py`
|-->`resources/`
|-->`assets/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`MyProgram/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`MyPyFile.py`

and I have `MyFunc()` inside `MyFile.py`, then how do I import `MyFunc()` in `app.py`? 

Normally, I could do `from assets.MyProgram.MyPyFile import MyFunc`, but if I do `briefcase dev`  then it gives error about not finding the `assets` directory. 
I know that I can use `self.paths.app` of `toga.App` object but not sure if it is the right way.

**Comment by freakboy3742** [comment, marked as answer]
You haven't got any `__init__.py` files on the `asset/MyProgram` path, so from Python's perspective, `assets` isn't on the PYTHONPATH. If you add `__init__.py` files, the path will become importable.

In addition: you say you have this structure "inside your src directory" - but you don't describe whether the path to `app.py` is `src/myapp/app.py` or `src/app.py`. If it's the former (which would be the default structure generated by `briefcase new`, the import path for `MyPyFile.py` will be `myapp.assets.MyProgram`, not `assets.MyProgram`.

**Reply by proneon267** [reply]
Yes you are right, the path is `src/myapp/app.py`. That worked! 
Also, if I now have:

|-->`assets/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`__init__.py`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`MyProgram/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`__init__.py`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`MyPyFile.py`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`bin/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`__init__.py`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|-->`Message.txt`

In `MyPyFile.py`, I have:
```
def Message():
	with open("./bin/Message.txt") as myfile:
		return myfile.read()
```
Now, if I do `briefcase dev` then I get error: `FileNotFoundError: [Errno 2] No such file or directory: './bin/Message.txt'`. 
I was thinking of using `self.paths.app` of `toga.App` object but that is in `app.py`. So, how should I reference the files in these cases?

**Reply by freakboy3742** [reply]
The root of the problem you're seeing is that `./` is a *relative* path. It depends on the current working directory. That's what `paths.app` allows you work around - it gives an absolute point in the filesystem to work with.

However, it's difficult to provide a full answer to your question because you're not giving the *entire* picture. Is `assets` in your `src/myapp` folder? Or is it a standalone folder in your project?

If it's in `src/myapp`, then `self.paths.app / "assets" / "bin" / "Message.txt"` should return the path you're looking for. If it's a standalone folder, then you'll probably need to use `self.paths.app.parent` as a root, and *also* add the assets folder to your `sources` definition so the extra directory will be copied into your project.

**Reply by proneon267** [reply]
Hello
Sorry for leaving out the complete directory structure:
```
.
├── LICENSE
├── logs
│   └── briefcase.2023_03_07-22_51_50.dev.log
├── pyproject.toml
├── README.rst
├── src
│   ├── myapp
│   │   ├── app.py
│   │   ├── assets
│   │   │   ├── __init__.py
│   │   │   ├── MyProgram
│   │   │   │   ├── bin
│   │   │   │   │   ├── __init__.py
│   │   │   │   │   └── Message.txt
│   │   │   │   ├── __init__.py
│   │   │   │   ├── MyPyFile.py
│   │   ├── __init__.py
│   │   ├── __main__.py
│   │   └── resources
│   │       ├── __init__.py
│   │       ├── myapp.icns
│   │       ├── myapp.ico
│   │       └── myapp.png
│   └── myapp.dist-info
│       ├── INSTALLER
│       ├── METADATA
│       ├── top_level.txt
│       └── WHEEL
└── tests
    ├── __init__.py
    ├── myapp.py
    └── test_app.py

13 directories, 29 files
```

`app.py`:
```
"""
My first application
"""
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW
from myapp.assets.MyProgram.MyPyFile import Message

class myapp(toga.App):
	
	def startup(self):
		"""
		Construct and show the Toga application.
		
		Usually, you would add your application to a main content box.
		We then create a main window (with a name matching the app), and
		show the main window.
		"""
		main_box = toga.Box()
		label = toga.Label(text=Message(self.paths.app))
		main_box.add(label)
		self.main_window = toga.MainWindow(title=self.formal_name)
		self.main_window.content = main_box
		self.main_window.show()


def main():
	return myapp()
```

`MyPyFile.py`:
```
import os
def Message(cwd):
	with open(os.path.join(cwd,"assets","MyProgram","bin","Message.txt")) as myfile:
		return myfile.read()
```

`Message.txt`:
```
Hello from Message.txt
```

The problem is I need `self.paths.app` in `MyPyFile.py` but it is accessible only in `app.py`--- where  `toga.App`  object is specified.
For this, I am passing `self.paths.app` as an argument to the `Message()` Function in `MyPyFile.py`. Is this the right way to do this?

**Reply by freakboy3742** [reply]
It's entirely valid to pass around `self.paths.app`; it's a Pathlib object, so it's a completely legitimate value to use as an argument. The only thing I can see that is unusual is the use of `os.path.join`; idiomatic use of Pathlib would be to say:
```
with (cwd / "assets" / "MyProgram" / "bin" / "Message.txt").open as myfile:
```
rather than using `os.path.join`; but otherwise, I can't see any reason it wouldn't work.

**Reply by proneon267** [reply]
Thanks for helping :)
